pkg/graveler/graveler.go

```diff
@@ -2528,6 +2530,124 @@ func (g *Graveler) Merge(ctx context.Context, repository *RepositoryRecord, dest
 	return commitID, nil
 }
 
+func (g *Graveler) MergeFromMetaRange(ctx context.Context, repository *RepositoryRecord, destination BranchID, source MetaRangeID, commitParams CommitParams) (CommitID, error) {
+	var (
+		preRunID string
+		commit   Commit
+		commitID CommitID
+	)
+
+	storageNamespace := repository.StorageNamespace
+	if err := g.prepareForCommitIDUpdate(ctx, repository, destination); err != nil {
+		return "", err
+	}
+
+	var tokensToDrop []StagingToken
+	// No retries on any failure during the merge. If the branch changed, it's either that commit is in progress, commit occurred,
+	// or some other branch changing operation. If commit is in-progress, then staging area wasn't empty after we checked so not retrying is ok.
+	// If another commit/merge succeeded, then the user should decide whether to retry the merge.
+	err := g.retryBranchUpdate(ctx, repository, destination, func(branch *Branch) (*Branch, error) {
+		empty, err := g.isSealedEmpty(ctx, repository, branch)
+		if err != nil {
+			return nil, fmt.Errorf("check if staging empty: %w", err)
+		}
+		if !empty {
+			return nil, fmt.Errorf("%s: %w", destination, ErrDirtyBranch)
+		}
+		toCommit, err := g.dereferenceCommit(ctx, repository, Ref(destination))
+		if err != nil {
+			return nil, err
+		}
+
+		// Source is not a commit - merge base is destination root
+		itr, err := g.Log(ctx, repository, toCommit.CommitID, false)
+		if err != nil {
+			return nil, err
+		}
+		root := toCommit
+		for itr.Next() {
+			root = itr.Value()
+		}
```

Comment:
Seeking root commit each time we merge does not scale. So unless we are creating a bare repository - we should have an initial commit that we should cache in the repository level? 
Alternative - can we just give the merge operation a fake one?
We do not keep track on this reference as a merge parent (it will make find base merge a nightmare) so a having something that will return an empty content for the merge should do the same work.
Alertnative 2 - enable merge without merge that will assume that we like to add files from source to destination when there is no base commit.

Reply:
Enable merge without merge will require us to modify code all the way down the stack. Will require changes in graveler, committed manager, range and meta range iterators and basically rewriting the whole merge flow.
Keeping track of the repository initial commit was discussed in a different context and was dismissed mostly because it is not backwards compatible. We cannot provide a fake one - since there are validation and require in-depth refactoring for it to work.
The alternative in my opinion is - creating a dangling commit